src/librustc_trans/trans/mir/rvalue.rs

```diff
@@ -198,7 +198,67 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
+                            let llval = operand.immediate();
+                            let discr = adt::trans_get_discr(bcx, &*repr, llval, None);
+                            (discr, common::val_ty(discr), adt::is_discr_signed(&*repr))
+                        } else {
+                            (operand.immediate(), ll_t_in, operand.ty.is_signed())
+                        };
+
+                        let newval = match (r_t_in, r_t_out) {
+                            (Int(_), Int(_)) => {
+                                let srcsz = ll_t_in.int_width();
+                                let dstsz = ll_t_out.int_width();
+                                if srcsz == dstsz {
+                                    build::BitCast(bcx, llval, ll_t_out)
+                                } else if srcsz > dstsz {
+                                    build::Trunc(bcx, llval, ll_t_out)
+                                } else if signed {
+                                    build::SExt(bcx, llval, ll_t_out)
+                                } else {
+                                    build::ZExt(bcx, llval, ll_t_out)
+                                }
+                            }
+                            (Float, Float) => {
+                                let srcsz = ll_t_in.float_width();
+                                let dstsz = ll_t_out.float_width();
+                                if dstsz > srcsz {
+                                    build::FPExt(bcx, llval, ll_t_out)
+                                } else if srcsz > dstsz {
+                                    build::FPTrunc(bcx, llval, ll_t_out)
+                                } else {
+                                    llval
+                                }
+                            }
+                            (Ptr(_), Ptr(_)) | (FnPtr, Ptr(_)) | (RPtr(_), Ptr(_)) =>
+                                build::PointerCast(bcx, llval, ll_t_out),
+                            (Ptr(_), Int(_)) | (FnPtr, Int(_)) =>
+                                build::PtrToInt(bcx, llval, ll_t_out),
+                            (Int(_), Ptr(_)) =>
+                                build::IntToPtr(bcx, llval, ll_t_out),
+                            (Int(_), Float) if signed =>
+                                build::SIToFP(bcx, llval, ll_t_out),
+                            (Int(_), Float) =>
+                                build::UIToFP(bcx, llval, ll_t_out),
+                            (Float, Int(IntTy::I)) =>
+                                build::FPToSI(bcx, llval, ll_t_out),
+                            (Float, Int(_)) =>
+                                build::FPToUI(bcx, llval, ll_t_out),
+                            _ => bcx.ccx().sess().bug(
+                                &format!("unsupported cast: {:?} to {:?}", operand.ty, cast_ty)
+                            )
+                        };
+                        OperandValue::Immediate(newval)
```

Comment:
Not sure if assuming everything is immediate here is correct, but I can’t think of a case where it wouldn’t be.